ui/raidboss/data/05-shb/raid/e8s.js

```diff
@@ -317,30 +317,173 @@
-        ko: '사슬 대상자',
-        cn: '连线',
+      regex: Regexes.tether({ id: '006E' }),
+      condition: function(data, matches) {
+        switch (data.options.e8sLightRampantStrat) {
+        case 'sharingan':
+          data.rampant = data.rampant || {
+            chainCount: 0,
+            chains: {},
+          };
+
+          data.rampant.chains[matches.source] = data.rampant.chains[matches.source] || [];
+          data.rampant.chains[matches.source].push(matches.target);
+          data.rampant.chains[matches.target] = data.rampant.chains[matches.target] || [];
+          data.rampant.chains[matches.target].push(matches.source);
+
+          // Need to track chains this way instead of counting sources on the object
+          // because the trigger will fire for both the third and fourth chain, and
+          // the third firing will have incomplete data
+          ++data.rampant.chainCount;
+
+          return data.rampant.chainCount === 4;
+        }
+
+        return data.me == matches.source;
+      },
+      response: function(data, matches) {
+        switch (data.options.e8sLightRampantStrat) {
+        case 'sharingan': {
+          let jobPriority = (job) => {
+            return ([
+              'GNB', 'DRK', 'WAR', 'PLD', // Tank Northwest
+              'SCH', 'WHM', 'AST', // Healer Northeast
+              'MNK', 'DRG', 'NIN', 'SAM', // Melee Southwest
+              'BRD', 'MCH', 'DNC', 'BLM', 'SMN', 'RDM', // Ranged Southeast
+            ].indexOf(job));
+          };
+          let chains = Object.keys(data.rampant.chains).sort((a, b) => {
+            return jobPriority(data.party.jobName(a)) - jobPriority(data.party.jobName(b));
+          });
+
+          // now chains[0] = tank, chains[1] = healer, chains[2] = melee DPS, chains[3] ranged DPS
+          // spots = 0, 1, 2, 3 equivalent to nw, ne, se, sw
+
+          let tank = chains[0];
+
+          let ret;
+          let alertFor;
+
+          if (!data.rampant.chains[tank].includes(chains[2])) {
+            // Box formation (no link to SE), tank swaps with SE
+            ret = {
+              en: 'Chain: Tank SE',
+            };
+            alertFor = [tank, chains[2]];
+          } else if (!data.rampant.chains[tank].includes(chains[3])) {
+            // Hourglass formation (no link to SW), tank swaps with SW
+            ret = {
+              en: 'Chain: Tank SW',
+            };
+            alertFor = [tank, chains[3]];
+          } else {
+            // Bowtie formation, tank stays
+            ret = {
+              en: 'Chain: Tank NW',
+            };
+          }
+
+          // If we need to move, alert us, otherwise just show info
+          if (alertFor.includes(data.me)) {
+            return {
+              alertText: ret,
+            };
+          }
+          // This is intentionally showing for all players
+          // not just players with chains, for callout purposes
+          return {
+            infoText: ret,
```

Comment:
I worry that this could be confusing, because calling out "Chain: Tank NW" or something for a non-chain player might make an orb player thing that they were a chain player?

Would it make sense to have something like:
- if you have a chain, alert Chain on YOU: Go/Swap NW
- if you don't have a chain, info "Chain Tank: NW"

Reply:
I went with just alerting on the tank chain position because they're the position that's always going to move to adjust to bowtie formation if an adjustment is needed. Everyone else should already know where they go for this mechanic (healer is always northeast, etc).

I could adjust the output text to show the swap text for the DPS that gets chained, my concern is if the healer is the one doing the callouts for the fight they won't see the tank position notification.

Reply:
It's not just the tank that needs to move though, it's a dps who needs to know if they are staying or swapping as well as the tank.  I'll change my proposal to:

- if you have a chain, alert Chain on YOU: Go/Swap NW/NE/SE/SW
- everybody, info "Chain Tank: NW"

And maybe disable tts for the "chain on you" one so they don't overlap and you still get the chain tank callout.

Reply:
If I return both `infoText` and `alertText` from the same `response` function, do they both get shown?

Reply:
Yeah, they both get shown.  Only the alert noise gets played.  The tts will be taken from the alert unless specified explicitly.

Reply:
Does ayatori actually make the tanks do everything? 

Looking at the top [ayatori video](https://www.youtube.com/watch?v=pmYgfT3Dbno) they do all swaps by the southwest. We stuck the ranged there for our group so we always make them adjust (it's free real estate!).

![image](https://user-images.githubusercontent.com/2417229/80062633-0a1adb00-84e9-11ea-912a-8061d4341034.png)

(Also i'm crying in being rDPS and doing all the callouts...)

Reply:
The three different explanations I saw (toolbox link, text guide that I've lost, and YT video that is probably somewhere in my YT history) all had tank swapping with SW or SE DPS.

Reply:
Hmm, this seems like a sharingan/NA specific thing....  And i guess the strat config is named sharingan so I shouldn't complain too much. The two ayatori videos I just perused (in addition to the one I just linked) swap the DPS at SW with everyone to form the correct shape.

Reply:
So you're saying we need an NA and a JP variation? :shrug:

Reply:
Nah. I was just perusing this PR since I was hoping to make my life slightly easier for callouts haha. This does make me +1 quisquous's comment about keeping them separate and toggleable between orbs and chains.

Reply:
Do note that having the tank swap 100% means the healer never has to, increasing their uptime slightly. Since PLD no longer has cast times during any cooldown block, this is a strict gain over "hope your physical ranged gets it".

Reply:
We're running a rphys, double caster {RDM, SMN}, and the two of us casters have instant casts up in time for the bowtie movement so no need to hope for phys ranged.

Our healer can be sad about their ruin2 or dropped GCD if they get unlucky 🤣 .

Reply:
It wouldn't be that difficult to separate it out into settings like this:

Strat: [none, sharingan]
Who moves: [Tank, melee DPS]

But I'm going to keep moving forward with the tank moves strat for now.

What should the orb strat be called? `sharingan` as well?

Reply:
"Orbs Through Middle"?

Reply:
👍 on separating in the future.

Yeah I think I started seeing orb strat on opposite side more with `sharingan` than with `ayatori` (though we also do it since it simplifies the movement for which E/W cardinal you pull into), so it seems reasonable to also call it `sharingan`. If Xenosys(?) uptime bowtie didn't rotate orbs CW twice (you end up in the same sides as this strat) it would have been nice to call this `uptime` orbs...